### PR TITLE
[linting] Apply PEP8-compliant module-level dunder names.

### LINKED ISSUE #12179: [linting] Enforce module-level dunder naming style?
**Is your feature request related to a problem? Please describe.**
After some recent [discussion](https://github.com/sphinx-doc/sphinx/pull/12144#discussion_r1532115125), we landed on using [PEP8 style Module Level Dunder Names](https://peps.python.org/pep-0008/#module-level-dunder-names) for a module `__all__` declaration:

https://github.com/sphinx-doc/sphinx/blob/b7f708dc634732aa90978e1c141ba6bd2af7ee84/tests/utils.py#L1-L5

**Describe the solution you'd like**
There is a `module-level-line` plugin for `flake8` that could achieve this for us.  Although.. we are considering removing `flake8` (see #12163).

**Describe alternatives you've considered**
It would be nice if `ruff` reported this - currently it doesn't seem aware of `MLL004`.

**Additional context**
 - Inspired by a code review question at https://github.com/sphinx-doc/sphinx/pull/12178#discussion_r1535943497

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Feature

### Purpose
- Add linting to check that module-level elements in our code adhere to PEP8 guidance.

### Detail
- Add the [`module-level-lint`](https://github.com/hanhwanglim/module-level-lint) `flake8` plugin to our `lint` extra-dependencies set.

### Relates
- Relates to (but does not long-term resolve) #12179.

### CODE REVIEW

**Comment by picnixz** [comment]
Personally, I think we should add it manually, incrementally. We don't want to keep flake8 or introduce new flake8 dependencies =/

**Comment by jayaddison** [comment]
@picnixz that's reasonable, yep.  Perhaps this branch exists solely to find existing instances of this that we can fix.

**Comment by jayaddison** [comment]
Hm.  Something isn't right here - commit a959887da00069ca1cd40db31506df61e6d18523 was supposed to fail.  I've moved the PR back into draft status.

**Comment by jayaddison** [comment]
> Hm. Something isn't right here - commit [a959887](https://github.com/sphinx-doc/sphinx/commit/a959887da00069ca1cd40db31506df61e6d18523) was supposed to fail. I've moved the PR back into draft status.

That failed because, despite `module-level-lint` being [added](https://github.com/sphinx-doc/sphinx/commit/a959887da00069ca1cd40db31506df61e6d18523#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R85) to [the `[test]` extras in `pyproject.toml`](https://github.com/sphinx-doc/sphinx/blob/a959887da00069ca1cd40db31506df61e6d18523/pyproject.toml#L83-L92), our [GitHub Actions `lint.yml` workflow spec](https://github.com/sphinx-doc/sphinx/blob/b1548d0aa4f89cf7e808b34ba4203b32f0fc67d1/.github/workflows/lint.yml) _does not_ specify that those `[lint]` extras should be installed during the [`flake8` workflow job](https://github.com/sphinx-doc/sphinx/blob/b1548d0aa4f89cf7e808b34ba4203b32f0fc67d1/.github/workflows/lint.yml#L42-L56).  An install of [`flake8` is requested](https://github.com/sphinx-doc/sphinx/blob/b1548d0aa4f89cf7e808b34ba4203b32f0fc67d1/.github/workflows/lint.yml#L51-L54) (alongside/subsequent to an upgrade of `pip`), but `MLL004` is only reported by `flake8` when `module-level-lint` is installed and available as a plugin.

File: sphinx/theming.py [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 from __future__ import annotations
 
+__all__ = 'Theme', 'HTMLThemeFactory'
```
Comment (picnixz) [inline sphinx/theming.py]:
```suggestion
__all__ = ('Theme', 'HTMLThemeFactory')
```

File: tests/test_extensions/ext_napoleon_pep526_data_google.py [hunk 2]
```diff
@@ -1,8 +1,5 @@
 """Test module for napoleon PEP 526 compatibility with google style"""
 
-module_level_var: int = 99
```
Comment (picnixz) [inline tests/test_extensions/ext_napoleon_pep526_data_google.py]: Keep it.

File: tests/test_extensions/ext_napoleon_pep526_data_numpy.py [hunk 3]
```diff
@@ -1,8 +1,5 @@
 """Test module for napoleon PEP 526 compatibility with numpy style"""
 
-module_level_var: int = 99
```
Comment (picnixz) [inline tests/test_extensions/ext_napoleon_pep526_data_numpy.py]: Keep it (remove this file from the lint)

**Review by picnixz — CHANGES_REQUESTED** [review]

**Comment by jayaddison** [comment]
Hi @picnixz - I've removed the additional `flake8` plugin dependency here, so all of this is cleanup/refactor.  Please could you re-review?

**Review by picnixz — APPROVED** [review]

**Comment by jayaddison** [comment]
Thank you!

**Pull request merged**